src/cpp/src/multinomial_decoding.cpp

```diff
@@ -0,0 +1,262 @@
+            token_scores.resize(pos - token_scores.data());
+        }
+
+        // sample next token
+        apply_softmax_inplace(token_scores);
+        for (size_t i = 0; i < token_scores.size(); i++) {
+            logits[i] = token_scores[i].score;
+        }
+
+        std::discrete_distribution<> dist(logits, logits + token_scores.size());
+        return token_scores[dist(gen)];
+    }
+};
+}  // namespace
+
+namespace ov {
+namespace genai {
+
+ov::genai::EncodedResults multinominal_decoding(ov::InferRequest& m_model_runner,
+                                                ov::Tensor input_ids,
+                                                ov::Tensor attention_mask,
+                                                ov::genai::GenerationConfig config,
+                                                std::shared_ptr<ov::genai::StreamerBase> streamer) {
+    ov::Shape prompts_shape = input_ids.get_shape();
+    size_t batch_size = prompts_shape[0];
+
+    OPENVINO_ASSERT(batch_size == 1, "Only batch size = 1 supported for multinomial decoding");
+
+    size_t prompt_len = prompts_shape[1];
+
+    ov::genai::EncodedResults results;
+    results.scores.resize(batch_size, 0);
+    results.tokens.resize(batch_size);
+
+    // Initialize inputs
+    m_model_runner.set_tensor("input_ids", input_ids);
+    m_model_runner.set_tensor("attention_mask", attention_mask);
+
+    ov::Tensor position_ids = m_model_runner.get_tensor("position_ids");
+    position_ids.set_shape(input_ids.get_shape());
+    std::iota(position_ids.data<int64_t>(), position_ids.data<int64_t>() + position_ids.get_size(), 0);
+
+    // Input values are persistent between inference calls.
+    // That allows to set values, which aren't going to change, only once
+    m_model_runner.get_tensor("beam_idx").set_shape({batch_size});
+    m_model_runner.get_tensor("beam_idx").data<int32_t>()[0] = 0;
+
+    m_model_runner.infer();
+
+    auto logits_tensor = m_model_runner.get_tensor("logits");
+
+    int64_t sequence_offset = logits_tensor.get_shape().at(1) - 1;
+    size_t vocab_size = logits_tensor.get_shape().back();
+
+    float* logits = logits_tensor.data<float>() + sequence_offset * vocab_size;
+
+    const int64_t* input_ids_data = input_ids.data<const int64_t>();
+
+    std::vector<int64_t> tokens{input_ids_data, input_ids_data + input_ids.get_size()};
+
+    RandomSampling sampling{config};
+
+    TokenIdScore out_token = sampling.get_out_token(logits, vocab_size, tokens);
+
+    tokens.push_back(out_token.id);
+    results.tokens[0].push_back(out_token.id);
+    results.scores[0] += out_token.score;
+
+    if (streamer) {
+        streamer->put(out_token.id);
+    }
+
+    if (!config.ignore_eos && out_token.id == config.eos_token_id) {
+        return results;
+    }
+
+    m_model_runner.get_tensor("input_ids").set_shape({batch_size, 1});
+    m_model_runner.get_tensor("position_ids").set_shape({batch_size, 1});

```

Comment:
what If position_id is not supported? as I know, it's not available for all models.
E.g. check opt-125m

Reply:
thanks for noticing, i will skip setting postion_ids for such model in the next push